### PR TITLE
feat(apigatewayv2): http api - metric methods for api and stage

### LINKED ISSUE #10325: [apigateway] Metric Method for HttpApi
Add the Metric method to the HttpApi object. 

Seems like most L2 objects have a method called metric, that takes a metric_name parameter, and makes it easy to build alarms for the resource. 

### Use Case

I built a config object that represents an alarm, and I loop over a dict of them to create alarms for each of my objects, I've been able to re use this on ECS, RDS, SQS, etc but have to build out the metric myself for API Gateway. 

This is what I want to work:

```
        for key, alarm in props.get('alarms').items():
            monitor = api.metric(alarm.metric_name).create_alarm(
                self, f"Alarm{alarm.metric_name}",
                evaluation_periods=alarm.eval_periods,
                threshold=alarm.threshold,
                actions_enabled=alarm.enabled,
                alarm_description=alarm.description,
                alarm_name=f"{props['ou']}-{props['env']}-api-{alarm.name}",
                comparison_operator=alarm.comparison_operator,
                datapoints_to_alarm=alarm.datapoints,
                period=alarm.period,
                statistic=alarm.statistic,
                treat_missing_data=alarm.missing
            )
            for action in alarm.actions:
                monitor.add_alarm_action(action)
```

### Proposed Solution

This is my work around:

```
        for key, alarm in props.get('alarms').items():
            metric = cloudwatch.Metric(
                metric_name=alarm.metric_name,
                namespace='AWS/ApiGateway',
                dimensions={
                    'ApiId': api.http_api_id
                }
            )
            monitor = metric.create_alarm(
                self, f"Alarm{alarm.metric_name}",
                evaluation_periods=alarm.eval_periods,
                threshold=alarm.threshold,
                actions_enabled=alarm.enabled,
                alarm_description=alarm.description,
                alarm_name=f"{props['ou']}-{props['env']}-api-{alarm.name}",
                comparison_operator=alarm.comparison_operator,
                datapoints_to_alarm=alarm.datapoints,
                period=alarm.period,
                statistic=alarm.statistic,
                treat_missing_data=alarm.missing
            )
            for action in alarm.actions:
                monitor.add_alarm_action(action)
```

### Other

Ultimately this is just a minor inconvenience and not high priority, but thought it should be tracked. 

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### LINKED ISSUE #10726: [apigateway] Getter for restApiName of RestApi class
Currently `RestApi` class doesn't have getter of `restApiName`

### Use Case
It would be good to have a getter of `restApiName` of `RestApi` class

### Proposed Solution

Should be very simple to add.

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by cyuste**
fixes #10325  
closes #10726

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/rest-api-base.ts [hunk 1]
```diff
@@ -273,6 +273,13 @@ export abstract class RestApiBase extends Resource implements IRestApi {
     return this._domainName;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/rest-api-base.ts:273]: Please revert this file rename.

**Review by nija-at — CHANGES_REQUESTED** [review]
The issue you've pointed at - #10325 - is referring to apigatewayv2, and not this module. Please provide a proper description in your commit message on what you're looking to achieve here.

If you're looking to add helper metric method to the `RestApi` construct, this is not done via Cfn augmentation. Instead, it's done explicitly in the source code. See here for examples - https://github.com/aws/aws-cdk/search?q=public+metric

Please also note that there is an open PR on related changes here - https://github.com/aws/aws-cdk/pull/10667

File: packages/@aws-cdk/aws-apigateway/lib/rest-api-base.ts (hunk 1 above)
Comment (cyuste) [reply packages/@aws-cdk/aws-apigateway/lib/rest-api-base.ts:273]: Hi,
Yes, I was trying to create metrics for RestApi.
I'm a little bit confused here... When should metrics be declared in the source code and when via cfn augmentations? For apigatewayv2 and HTTP Api specifically, what would be the correct approach?
Thanks

Comment (nija-at) [reply packages/@aws-cdk/aws-apigateway/lib/rest-api-base.ts:273]: The correct approach would be to add these in code.

Comment (cyuste) [reply packages/@aws-cdk/aws-apigateway/lib/rest-api-base.ts]: Added to code

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts [hunk 2]
```diff
@@ -233,4 +241,75 @@ export class HttpApi extends Resource implements IHttpApi {
...
+  /**
+   * Return the given named metric for this HTTP Api Gateway
+   *
+   * @default Average over 5 minutes
+   */
+  public metric(metricName: string, props?: MetricOptions): Metric {
+    return new Metric({
+      namespace: 'AWS/ApiGateway',
+      metricName,
+      dimensions: { apiName: this.httpApiName },
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:254]: From what I see here - https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-metrics.html - apiName is not a dimension. It should be `apiId`.

If that's correct, also revert the changes for API name above.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts [hunk 3]
```diff
@@ -117,4 +119,73 @@ export class HttpStage extends Resource implements IStage {
...
     return `https://${this.httpApi.httpApiId}.execute-api.${s.region}.${s.urlSuffix}/${urlPath}`;
   }
+
+  /**
+   * Return the given named metric for this HTTP Api Gateway Stage
+   *
+   * @default Average over 5 minutes
+   */
+  public metric(metricName: string, props?: MetricOptions): Metric {
+    var api = this.httpApi as HttpApi;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts:129]: This is not guaranteed to be true. It can also be an imported HttpApi, in which case, this will fail with an error.

It's safe to elevate the `metric*()` APIs to the `IHttpApi` interface and introducing an `HttpApiBase` class. This way the logic will work for both imported and CDK implemented `IHttpApi`.

Change the code like so -

```ts
export interface IHttpApi extends IResource {
  // ... what's already there

  metric(metricName: string, props?: MetricOptions): Metric;

  metricClientError(props?: MetricOptions): Metric;

  // ...
}

class HttpApiBase extends Resource implements IHttpApi { // note that this is not exported
  public metric(metricName: string, props?: MetricOptions): Metric {
    // ...
  }

  public metricClientError(props?: MetricOptions): Metric {
    // ...
  }
}

export class HttpApi extends HttpApiBase {
  public static fromApiId(scope: Construct, id: string, httpApiId: string): IHttpApi {
    class Import extends HttpApiBase {
      public readonly httpApiId = httpApiId;
    }
    return new Import(scope, id);
  }

  // ...
}
```

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for making the changes. My comments are below.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts [hunk 4]
```diff
@@ -233,4 +241,75 @@ export class HttpApi extends Resource implements IHttpApi {
...
+    return this.metric('5XXError', { statistic: 'Sum', ...props });
+  }
+
+  /**
+   * Metric for the amount of data processed in bytes.
+   *
+   * @default - sum over 5 minutes
+   */
+  public metricDataProcessed(props?: MetricOptions): Metric {
+    return this.metric('CacheHitCount', { statistic: 'Sum', ...props });
```
[the remark this answers is not in the document]
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts]: This was a stupid mistake, fixed

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts (hunk 2 above)
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts]: I have left httpApiName property and fixed dimensions

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts (hunk 3 above)
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts]: HttpApiBase created and metrics moved to interface.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts [hunk 5]
```diff
@@ -16,6 +17,12 @@ export interface IHttpApi extends IResource {
    */
   readonly httpApiId: string;
 
+  /**
+   * A human friendly name for this HTTP API. Note that this is different from `httpApiId`.
+   * @attribute
```
[the remark this answers is not in the document]
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts]: Removed attribute tag

File: packages/@aws-cdk/aws-apigatewayv2/README.md [hunk 6]
```diff
@@ -198,3 +199,22 @@ with 3 API mapping resources across different APIs and Stages.
...
+## Metrics
+
+The API Gateway v2 service sends metrics around the performance of HTTP APIs to Amazon CloudWatch. These metrics can be referred to using the metric APIs available on the HttpApi construct. The APIs with the metric prefix can be used to get reference to specific metrics for this API. For example, the method below refers to the client side errors metric for this API.
+
+```
+const api = new apigw.HttpApi(stack, 'my-api');
+const clientErrorMetric = api.metricClientError();
+
+```
+Please note that this will return a metric for all the stages defined in the api. It is also possible to collect metrics only from a specic `stage` using the metric methods from the `stage`.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/README.md:212]: Slight adjustments to formatting

````suggestion
```

Please note that this will return a metric for all the stages defined in the api. It is also possible to refer to metrics for a specific Stage using
the `metric` methods from the `Stage` construct.
````

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts [hunk 7]
```diff
@@ -16,10 +17,67 @@ export interface IHttpApi extends IResource {
    */
   readonly httpApiId: string;
 
+  /**
+   * A human friendly name for this HTTP API. Note that this is different from `httpApiId`.
+   */
+  readonly httpApiName?: string;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:23]: Since this is only available to CDK initialized http api (not imported), remove this from the interface and declare it only in `HttpApi`.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts [hunk 8]
```diff
@@ -140,7 +187,7 @@ export class HttpApi extends Resource implements IHttpApi {
   constructor(scope: Construct, id: string, props?: HttpApiProps) {
     super(scope, id);
 
-    const apiName = props?.apiName ?? id;
+    this.httpApiName = props?.apiName ?? id;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:191]: This entire constructor needs to move back into `HttpApi`. It's only applicable when users create a new `HttpApi` in the CDK. 

You should be able to declare `HttpApiBase` as `abstract` and `httpApiId` property as `abstract`, and leave the class without a constructor.
The abstract `httpApiId` will then have to be initialized in the constructors of `HttpApi` and `Import`.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts [hunk 9]
```diff
@@ -117,4 +119,73 @@ export class HttpStage extends Resource implements IStage {
     const urlPath = this.stageName === DEFAULT_STAGE_NAME ? '' : this.stageName;
     return `https://${this.httpApi.httpApiId}.execute-api.${s.region}.${s.urlSuffix}/${urlPath}`;
   }
+
+  /**
+   * Return the given named metric for this HTTP Api Gateway Stage
+   *
+   * @default Average over 5 minutes
+   */
+  public metric(metricName: string, props?: MetricOptions): Metric {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts:128]: You could apply the same logic as `HttpApi` of moving this into the interface and creating an abstract `HttpStageBase` class.

This is entirely optional if you want to skip it.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts [hunk 10]
```diff
@@ -16,10 +17,67 @@ export interface IHttpApi extends IResource {
...
+
   /**
    * The default stage
    */
   readonly defaultStage?: HttpStage;
+
+  /**
+   * Return the given named metric for this HTTP Api Gateway
+   *
+   * @default Average over 5 minutes
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:33]:
```suggestion
   * @default - average over 5 minutes
```

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts [hunk 11]
```diff
@@ -117,4 +119,73 @@ export class HttpStage extends Resource implements IStage {
     const urlPath = this.stageName === DEFAULT_STAGE_NAME ? '' : this.stageName;
     return `https://${this.httpApi.httpApiId}.execute-api.${s.region}.${s.urlSuffix}/${urlPath}`;
   }
+
+  /**
+   * Return the given named metric for this HTTP Api Gateway Stage
+   *
+   * @default Average over 5 minutes
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts:126]:
```suggestion
   * @default - average over 5 minutes
```

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for making the suggested changes. 
Just one major comment around where the constructor should go. See below.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts (hunk 8 above)
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:241]: done

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts (hunk 7 above)
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts]: done

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts (hunk 9 above)
Comment (cyuste) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts]: As you prefer. Currently I can import an api and get its metrics (created a new test for it) but I can't for a stage, what is correct as I cannot either associate an imported stage to an api and therefore cannot set the metric dimensions.

Comment (nija-at) [reply packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts]: It's fine. It can be done later as a separate PR.

**Review by nija-at — APPROVED** [review]
This is looking great now!
